src/main/java/org/jboss/pnc/common/http/PNCHttpClient.java

```diff
@@ -0,0 +1,181 @@
+                .withBackoff(retryConfig.backoffInitialDelay(), retryConfig.backoffMaxDelay(), ChronoUnit.SECONDS)
+                .withMaxRetries(retryConfig.maxRetries())
+                .withMaxDuration(Duration.of(retryConfig.maxDuration(), ChronoUnit.SECONDS))
+                .onSuccess(e -> log.debug("Request successfully sent, response: {}", e.getResult().statusCode()))
+                .onRetry(
+                        e -> log.debug(
+                                "Retrying (attempt #{}), last response was.",
+                                e.getAttemptCount(),
+                                e.getLastException()))
+                .onFailure(e -> log.warn("Request couldn't be sent.", e.getException()))
+                .build();
+    }
+
+    /**
+     * Tries to send the request once, reporting any error immediately.
+     *
+     * @param request Request to be sent.
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void trySendRequest(Request request) {
+        trySendRequest(request, request.getAttachment());
+    }
+
+    /**
+     * Tries to send the request once replacing the payload from {@link Request#getAttachment()} with provided payload,
+     * reporting any error immediately.
+     *
+     * @param request Request to be sent.
+     * @param payload Payload to be used instead of {@link Request#getAttachment()}
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void trySendRequest(Request request, Object payload) {
+        HttpRequest httpRequest = prepareHttpRequest(request, payload);
+        doSendRequest(httpRequest);
+    }
+
+    /**
+     * Sends the request, retrying according to the retry config if there is any error.
+     *
+     * @param request Request to be sent.
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void sendRequest(Request request) {
+        sendRequest(request, request.getAttachment());
+    }
+
+    /**
+     * Send the request replacing the payload from {@link Request#getAttachment()} with provided payload, retrying
+     * according to the retry config if there is any error.
+     *
+     * @param request Request to be sent.
+     * @param payload Payload to be used instead of {@link Request#getAttachment()}
+     * @throws PNCHttpClientException When there is error sending the request.
+     */
+    public void sendRequest(Request request, Object payload) {
+        HttpRequest httpRequest = prepareHttpRequest(request, payload);
+        Failsafe.with(retryPolicy).run(() -> doSendRequest(httpRequest));
+    }
+
+    private void doSendRequest(HttpRequest httpRequest) {
+        log.debug("Sending the request {}", httpRequest);
+        try {
+            HttpResponse<Void> response = client.send(httpRequest, HttpResponse.BodyHandlers.discarding());
+            if (response.statusCode() >= 300) {
```

Comment:
I don't think retrying on redirect helps. I'd add it an option to follow redirects for the http client?

Reply:
There is an option to follow redirects already. I just want to fail if I end up with any non 2xx status